xarray/tests/test_dataarray.py

```diff
@@ -1023,32 +1024,45 @@ def test_sel_dataarray_datetime_slice(self) -> None:
         result = array.sel(delta=slice(array.delta[0], array.delta[-1]))
         assert_equal(result, array)
 
-    def test_sel_float(self) -> None:
+    @pytest.mark.parametrize(
+        ["coord_values", "indices"],
+        (
+            pytest.param(
+                np.array([0.0, 0.111, 0.222, 0.333], dtype="float64"),
+                slice(1, 3),
+                id="float64",
+            ),
+            pytest.param(
+                np.array([0.0, 0.111, 0.222, 0.333], dtype="float32"),
+                slice(1, 3),
+                id="float32",
+            ),
+            pytest.param(
+                np.array([0.0, 0.111, 0.222, 0.333], dtype="float16"),
+                slice(1, 3),
+                id="float16",
+                marks=[
+                    pytest.mark.skipif(
+                        has_pandas_version_two, reason="not supported for pandas >= 2.0"
+                    )
+                ],
+            ),
```

Comment:
this is the main change here: we skip the `float16` check with `pandas>=2.0`. Another option would be to change the code to explicitly cast to `float32` / `float64`, then remove the `skipif`

Reply:
explicit cast sounds good to me.

Reply:
to which precision should we cast? `float64`?

Reply:
Yes, IIUC these arrays are converted to pandas Indexes, which used to upcast to `float64` always.

Perhaps we should upcast for backwards compatibility and raise a `DeprecationWarning` asking the user to cast explicitly

Reply:
this turns out not to be as simple as I thought it would be: the cast to `PandasIndex` (and thus `pandas.Index`) happens in multiple places. First is the `DataArray` / `Dataset` constructor, which through several layers calls `safe_cast_to_index`, which is where the actual call to `PandasIndex` happens. The second cast happens when selecting using a array of values. In that case, `get_indexer_nd` calls `index.get_indexer` (or rather, `pandas.Index.get_indexer`), which appears to create a index from the indexer values.

Should both casts emit the `DeprecationWarning`? I think the second might be a implementation detail and that we should just cast the indexer values from `float16` to `float64` (not sure, though).